cpp/include/cuspatial/distances/linestring_distance.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <memory>
+
+namespace cuspatial {
+
+/**
+ * @brief Compute shortest distance between pairs of linestrings (a.k.a. polylines)
+ *
+ * The shortest distance between two linestrings is defined as the shortest distance
+ * between all pairs of segments of the two linestrings. If any of the segments intersect,
+ * the distance is 0.
+ *
+ * The following example contains 4 pairs of linestrings.
+ *
+ * First pair:
+ * (0, 1) -> (1, 0) -> (-1, 0)
+ * (1, 1) -> (2, 1) -> (2, 0) -> (3, 0)
+ *
+ *     |
+ *     *   #---#
+ *     | \     |
+ * ----O---*---#---#
+ *     | /
+ *     *
+ *     |
+ *
+ * The shortest distance between the two linstrings are the distance
+ * from point (1, 1) to segment (0, 1) -> (1, 0), which is sqrt(2)/2.
+ *
+ * Second pair:
+ *
+ * (0, 0) -> (0, 1)
+ * (0, 2) -> (0, 3) -> (1, 3)
+ *
+ * Both linestrings contains degenerate segments, their distance is 1.
+ *
+ * Third pair:
+ *
+ * (0, 0) -> (2, 2) -> (-2, 0)
+ * (2, 0) -> (0, 2)
+ *
+ * These linestrings intersects, their distance is 0
+ *
+ * Forth pair:
+ *
+ * (2, 2) -> (-2, -2)
+ * (1, 1) -> (5, 5) -> (10, 0)
+ *
+ * These linestrings contain collinear and overlapping sections,
```

Comment:
```suggestion
 * These linestrings contain colinear and overlapping sections, so
```

Reply:
I've always remembered `colinear` with 2 `l`s. Updating the method name in `linestring_distance.cu` too.

Reply:
Actually both spelling works: https://www.merriam-webster.com/dictionary/colinear

Reply:
Yeah TIL they both work. Pick one. I have always liked colinear because it's the prefix co plus the root linear. It matches coplanar. But it seems many people use collinear. English is tricky that way.